pkg/csi/service/wcp/controller.go

```diff
@@ -462,18 +462,27 @@ func (c *controller) createBlockVolume(ctx context.Context, req *csi.CreateVolum
 	topologyRequirement = req.GetAccessibilityRequirements()
 	filterSuspendedDatastores := commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.CnsMgrSuspendCreateVolume)
 	isTKGSHAEnabled := commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.TKGsHA)
+	isVdppOnStretchedSVEnabled := commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.VdppOnStretchedSupervisor)
 	topoSegToDatastoresMap := make(map[string][]*cnsvsphere.DatastoreInfo)
 	if isTKGSHAEnabled {
 		// TKGS-HA feature is enabled
-		// Identify the topology keys in Accessibility requirements.
+		// Identify the topology keys in Accessibility requirements and infer the environment type based on these keys.
+
+		// If both zone and hostname labels are present in the request -
+		// it's host local volume provisioning on a stretched supervisor.
+
+		// If only zone label is present in the request -
+		// it's non-host local volume provisioning on a stretched supervisor.
+
+		// If only hostname label is present in the request -
+		// it's host local volume provisioning on a non-stretched supervisor.
+
+		// If neither zone nor hostname label is present in the request -
+		// it's non-host local volume provisioning on a non-stretched supervisor.
 		hostnameLabelPresent, zoneLabelPresent = checkTopologyKeysFromAccessibilityReqs(topologyRequirement)
-		// TODO: TKGS-HA: This case will only arise when spherelet will add zone and hostname labels to CSINodes.
-		// Currently spherelet only accepts hostname. We will handle this case later.
-		if zoneLabelPresent && hostnameLabelPresent {
-			return nil, csifault.CSIUnimplementedFault, logger.LogNewErrorCodef(log, codes.Unimplemented,
-				"support for topology requirement with both zone and hostname labels is not yet implemented.")
+		if isVdppOnStretchedSVEnabled && hostnameLabelPresent {
+			log.Infof("Host Local volume provisioning on stretched supervisor with requirement: %+v", topologyRequirement)
 		} else if zoneLabelPresent {
```

Comment:
what about when isVdppOnStretchedSVEnabled FSS is enabled, hostnameLabelPresent is true and zoneLabelPresent is true?

treat FSS isVdppOnStretchedSVEnabled as always true post 9.0 regardless of deployment is single cluster or stretched cluster.

Reply:
Ideally try to restructure the code in the following way to be clear:
```
if isTKGSHAEnabled {
    if isVdppOnStretchedSVEnabled {
        if hostnameLabelPresent {
        // Code for vdpp provisioning - stretched & non-stretched SV
        } else if zoneLabelPresent {
        // Code for zonal volume provisioning
        } else {
        // Code for volume provisioning for non-vdpp workloads in single supervisor
        }
    } else {
    // Older code
    }
} 
```

Reply:
@divyenpatel if the FSS is enabled and both host and zone labels are present then we will go with host local volume provisioning. 

@shalini-b IMO it's already getting harder to understand with too many if-else nests. Let's keep it this way for now and we can do refactoring as part of a separate task.